formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonNamesMap.kt

```diff
@@ -11,46 +12,71 @@ import kotlinx.serialization.json.*
 import kotlin.native.concurrent.*
 
 @SharedImmutable
-internal val JsonAlternativeNamesKey = DescriptorSchemaCache.Key<Map<String, Int>>()
+internal val JsonDeserializationNamesKey = DescriptorSchemaCache.Key<Map<String, Int>>()
 
-@OptIn(ExperimentalSerializationApi::class)
-internal fun SerialDescriptor.buildAlternativeNamesMap(): Map<String, Int> {
+@SharedImmutable
+internal val JsonSerializationNamesKey = DescriptorSchemaCache.Key<Array<String>>()
+
+internal fun SerialDescriptor.buildDeserializationNamesMap(json: Json): Map<String, Int> {
     fun MutableMap<String, Int>.putOrThrow(name: String, index: Int) {
         if (name in this) {
             throw JsonException(
                 "The suggested name '$name' for property ${getElementName(index)} is already one of the names for property " +
-                        "${getElementName(getValue(name))} in ${this@buildAlternativeNamesMap}"
+                        "${getElementName(getValue(name))} in ${this@buildDeserializationNamesMap}"
             )
         }
         this[name] = index
     }
 
-    var builder: MutableMap<String, Int>? = null
+    val builder: MutableMap<String, Int> =
+        mutableMapOf() // can be not concurrent because it is only read after creation and safely published to concurrent map
+    val strategy = namingStrategy(json)
     for (i in 0 until elementsCount) {
         getElementAnnotations(i).filterIsInstance<JsonNames>().singleOrNull()?.names?.forEach { name ->
-            if (builder == null) builder = createMapForCache(elementsCount)
-            builder!!.putOrThrow(name, i)
+            builder.putOrThrow(name, i)
         }
+        strategy?.let { builder.putOrThrow(it.serialNameForJson(this, i, getElementName(i)), i) }
     }
-    return builder ?: emptyMap()
+    return builder.ifEmpty { emptyMap() }
 }
 
+internal fun Json.deserializationNamesMap(descriptor: SerialDescriptor): Map<String, Int> =
```

Comment:
There may be it worth in the name or comment to explicitly indicate that these are alternative names (not those that match the name of the property)?